basis/compiler/cfg/builder/alien/boxing/boxing.factor

```diff
@@ -10,19 +10,33 @@ IN: compiler.cfg.builder.alien.boxing
 
 SYMBOL: struct-return-area
 
+SYMBOLS: int-reg-reps float-reg-reps ;
+
+: record-reg-reps ( reps -- reps )
+    dup  ! reps: { { reg-rep on-stack? odd-register? } ... }
+    [ [ [ first int-rep?     ] [ second not ] bi and ] count int-reg-reps   +@ ]
```

Comment:
Maybe these are simpler like this:

```factor
: record-reg-reps ( reps -- reps )
    dup [
        dup second [ drop ] [ first int-rep? int-reg-reps float-reg-reps ? inc ] if
    ] each ;

: unrecord-reg-reps ( reps -- reps )
    dup [
        dup second [ drop ] [ first int-rep? int-reg-reps float-reg-reps ? dec ] if
    ] each ;
```

Or maybe (although this has an intermediate array created):

```factor
: record-reg-reps ( reps -- reps )
    dup [ second ] reject [ first int-rep? int-reg-reps float-reg-reps ? inc ] each ;

: unrecord-reg-reps ( reps -- reps )
    dup [ second ] reject [ first int-rep? int-reg-reps float-reg-reps ? dec ] each ;
```

At the least, ``-1 *`` should be ``neg``, but it would be nice to iterate once, or simplify these similar expressions below.

Reply:
Or even, doing a single namespace increment:

```factor
: reg-reps ( reps -- int-reps float-reps ):
    [ second ] reject [ [ first int-rep? ] count ] [ length over - ] bi

: record-reg-reps ( reps -- reps )
    dup reg-reps int-reg-reps float-reg-reps [ +@ ] bi-curry@ bi* ;

: unrecord-reg-reps ( reps -- reps )
    dup reg-reps [ neg ] bi@ int-reg-reps float-reg-reps [ +@ ] bi-curry@ bi* ;
```

Reply:
On a related question, is it true that these are either "int-rep" or "float-rep" for the purposes of your change?  Or at least that practically that's fine when you have other c-types in a struct?

Reply:
> On a related question, is it true that these are either "int-rep" or "float-rep" for the purposes of your change? Or at least that practically that's fine when you have other c-types in a struct?

The names of the variables `int-reg-reps` and `float-reg-reps` are directly related to CPU register types, rather than C type names.
For example, in X86-64:

```Factor
IN: cpu.x86.64.unix

M: x86.64 param-regs
    drop {
        { int-regs { RDI RSI RDX RCX R8 R9 } }
        { float-regs { XMM0 XMM1 XMM2 XMM3 XMM4 XMM5 XMM6 XMM7 } }
    } ;
```
The word that Factor uses to allocate the actual registers for its arguments is `next-reg-param` :
```Factor
IN: compiler.cfg.builder.alien.params
GENERIC: next-reg-param ( odd-register? rep -- reg )

USING: compiler.cfg.builder.alien.params cpu.architecture kernel
math namespaces sequences ;
M: int-rep next-reg-param
    [ nip ?dummy-stack-params ] [ nip ?dummy-fp-params ] [
        drop
        [ int-regs get last even? [ int-regs get pop* ] when ]
        when
    ] 2tri int-regs get pop ;

USING: compiler.cfg.builder.alien.params cpu.architecture kernel
namespaces sequences ;
M: object next-reg-param
    nip [ ?dummy-stack-params ] [ ?dummy-int-params ] bi
    float-regs get pop ;
```

Here, `int-rep` is allocated from the registers of `int-regs`, while others are allocated from `float-regs`.

Factor provides `int-rep` for C type arguments that use integer registers.

```Factor
IN: scratchpad USING: alien.c-types compiler.cfg.builder.alien.boxing classes.struct ;
IN: scratchpad int base-type flatten-c-type .
{ { int-rep f f } }
alien.c-types:short base-type flatten-c-type .
{ { int-rep f f } }
IN: scratchpad long base-type flatten-c-type .
{ { int-rep f f } }
IN: scratchpad STRUCT: s1 { m1 long } { m2 ulong } ;
IN: scratchpad s1 base-type flatten-c-type .
{ { int-rep f f } { int-rep f f } }

Reply:
I think all three suggestions are better than the original. The third, in particular, allows us to use `reg-reps` elsewhere, making the code cleaner. 
However, the name may indicate that `reg-reps` does not include the reps of `on-stack?`.